SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/Storage.cs

```diff
@@ -369,18 +367,40 @@ private Task<T> AddTaskToQueue<T>(Func<T> action)
                 {
                     return action();
                 }
-                catch (Exception e) when (!(e is StorageException))
+                catch (Exception e)
                 {
-                    // Tasks should already be throwing only storage exceptions, but in case any are missed, 
-                    // which has happened, catch them here and wrap in a storage exception. This will prevent 
-                    // the exception from being unobserved.
-                    throw new StorageException(e);
+                    // And regarding the comment in other variant of this function, async lambda + await cannot work with a Func anyway.
+                    throw HandleStorageRelatedExceptionAsync(e).GetAwaiter().GetResult();
                 }
             });
             AddTaskToQueue(task);
             return task;
         }
 
+        private async Task<Exception> HandleStorageRelatedExceptionAsync(Exception e)
+        {
+            // Check if database is corrupted, we have evidence (https://github.com/microsoft/appcenter-sdk-dotnet/issues/1184)
+            // that it's not always originated by a proper SQLiteException (which would then be converted to StorageException in StorageAdapter).
+            // If it was always the right type then the exception would not have been unobserved in that application before we changed the re-throw logic here.
+            // But the message is definitely "Corrupt" and thus unfortunately that is the only check we seem to be able to do as opposed to type/property checking.
+            if (e.Message == "Corrupt")
```

Comment:
Do we sure that case-sensitive comparison is enough? Maybe we should use "contains"?

Reply:
Right now all occurrences I found while searching for this is this specific casing.
Now the check needs to be improved as it could be a legit inner SQLiteException, so need to check inner on 1 level as well.